pkg/controller/pravegacluster/upgrade.go

```diff
@@ -404,6 +405,139 @@ func (r *ReconcilePravegaCluster) syncSegmentStoreVersion(p *pravegav1alpha1.Pra
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: newsts.Name, Namespace: p.Namespace}, newsts)
+	//this check is to see if the newsts is present or not if it's not present it will be created here
+	if err != nil && errors.IsNotFound(err) {
+		*newsts.Spec.Replicas = 0
+		err = r.client.Create(context.TODO(), newsts)
+		if err != nil {
+			return false, err
+		}
+	}
+	//here we are getting the name of the oldsts based on either we are upgrading or doing rollback
+	var name string = ""
+	oldsts := &appsv1.StatefulSet{}
+	if util.IsClusterUpgradingTo07(p) {
+		name = util.StatefulSetNameForSegmentstoreBelow07(p.Name)
+	} else {
+		name = util.StatefulSetNameForSegmentstoreAbove07(p.Name)
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: p.Namespace}, oldsts)
+	//this check is to see if the old sts is present or not
+	if err != nil && errors.IsNotFound(err) {
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: newsts.Name, Namespace: p.Namespace}, newsts)
+		//this checks is to verify if the old sts is not present new sts should be present
+		if err != nil && errors.IsNotFound(err) {
+			return false, err
+		}
+		//this is the condition where we are checking if the new ready ss pods are equal to the ss replicas thus meaning upgrade is completed
+		if newsts.Status.ReadyReplicas == p.Spec.Pravega.SegmentStoreReplicas {
+			return true, nil
+		}
+	}
+
+	//this is to check if some pod has come into an error state or not
+	if oldsts.Status.ReadyReplicas+newsts.Status.ReadyReplicas != p.Spec.Pravega.SegmentStoreReplicas {
+		//this will get all the pods created with this target version till now
+		pods, err := r.getStsPodsWithVersion(newsts, p.Status.TargetVersion)
+		if err != nil {
+			return false, err
+		}
+		//checking if any of above pods have gone into error sate
+		_, err = r.checkUpdatedPods(pods, p.Status.TargetVersion)
+		if err != nil {
+			// Abort if there is any errors with the updated pods
+			return false, fmt.Errorf("updating statefulset (%s) failed due to %v", newsts.Name, err)
+		}
+	}
+
+	p.Status.UpdateProgress(pravegav1alpha1.UpdatingSegmentstoreReason, "0")
+	//this check to ensure that the oldsts always decrease by 2 as well as newsts pods increase by 2 only then the next increment or decrement happen
+	//first condition is true and used in the case of rollbackCondition
+	//second condition is true and used in case of Upgrade
+	if (r.IsClusterRollbackingFrom07(p) && newsts.Status.ReadyReplicas == *newsts.Spec.Replicas) || (oldsts.Status.ReadyReplicas+newsts.Status.ReadyReplicas == p.Spec.Pravega.SegmentStoreReplicas && newsts.Status.ReadyReplicas == *newsts.Spec.Replicas) {
+		//this check is run till the value of old sts replicas is greater than 0 and will increase two replicas of the new sts and delete 2 replicas of the old sts
+		if *oldsts.Spec.Replicas > 2 {
+			*newsts.Spec.Replicas = *newsts.Spec.Replicas + 2
```

Comment:
Encapsulate the changes to newSTS in a function and invoke that here.
Ditto for old STS, so its easier to follow what is happening.

Reply:
done